### ISSUE TITLE
Environment Matching helper function

### ISSUE THREAD

**Issue by bdice** [glotzerlab/freud#278]
As suggested by @bvansade, it could be useful to create a helper function that assists users in selecting parameters for `MatchEnv`. In my experience, it's rather difficult to find settings that work well without a deep understanding of the module and its algorithms, so any sort of helpful heuristic would be nice. Perhaps this would work by scanning a range of settings and doing some kind of optimization.

@erteich @jglaser @xiyudu I'd be happy to hear your thoughts on this!

**Comment by xiyudu** [comment]
I haven't used the MatchEnv module ever since I wrote some unit test for it long long ago.  So I actually never used the module to process "real" data.  However, for my current project, I might start using the module really soon, so maybe I will have some helpful insight later on in the summer.  But I will be happy to be a part of the discussion.

**Comment by bdice** [comment]
@xiyudu Sorry, I thought you'd used it for one of your projects! Happy to have your thoughts and I appreciate your willingness to offer feedback! I know the other two I tagged have used it productively for a thing or two. 😄

**Comment by erteich** [comment]
Hi all, I'd be happy to help with this! Do you mean the threshold specifically? My rule of thumb, that usually works quite well, is to choose a threshold of 0.2, and a cutoff radius of the first minimum in g(r), so the nearest neighbor shell. I just took a look at the "cluster" documentation and this isn't clear, perhaps I should update it to better clarify, two environment vectors are deemed "matching" if the distance between them falls below threshold*cutoff_radius. This way, the threshold is dimensionless. 

I'd be happy to think about parameter sweeps to implement; were there other methods besides cluster() that you were especially focused on?

**Comment by bdice** [comment]
@erteich Yes, the essential goal is to make the "cluster" algorithm work more easily. It's pretty heavily dependent on finding the exact right set of parameters right now, and I never know how to help people make it work for their systems - I haven't been able to build an intuition for it as you have. Your descriptions would be helpful in the docs, for sure!

**Comment by vyasr** [comment]
I think it would also be nice to have some guidance on when a user would expect to use which method, and which fields get populated by each method.

**Comment by vyasr** [comment]
Resolved in #489.

**Comment by Charlottez112** [comment]
I'm reopening this issue because I have the same experience with the `Environment` module. I think it is very useful, and I really want to use it more, but currently it's kinda hard to figure out what parameters one should use. (more specifically `EnvironmentCluster` )

`threshold` is definitely a parameter that needs tuning, but sometimes the user might not know what kind of neighbor list to use (`r_max` vs `num_neighbors`, and how to determine these parameters). I'm not entirely sure what `neighbors` in the arguments mean. Based on the documentation, it seems like this should always be greater than or equal to `env_neighbors`, but I found setting `neighbors` a smaller number than `env_neighbors` wouldn't make it error.

My user case: I'm testing it out on a perfect gamma Brass lattice (cI52 Cu5Zn8), which has 4 Wyckoff sites in one unit cell. I thought I could get back the environment clusters fairly easily since this is a perfect lattice with point particles, and I'm expecting 1 or 4 clusters (1 because it's one structure or grain, 4 because of the 4 Wyckoff sites), but it would give me much more clusters (~100 at least for a lattice with about 400 points), and I tried a lot of combinations of `neighbors` and `env_neighbors`. Still I'll need to visualize the results to see if the cluster assignment makes sense, and to understand why there are so many particles that are labeled as different clusters even for a perfect crystal, because I don't know a better way to check the results.

I feel like it would be really helpful if we can have a tutorial that shows step by step how one can get the environment clusters for a slightly complex perfect crystal. By "step" I mean how to decide what parameters to use or how to guess to narrow down the ranges we should try.

**Comment by bdice** [comment]
> I'm not entirely sure what `neighbors` in the arguments mean. Based on the documentation, it seems like this should always be greater than or equal to `env_neighbors`, but I found setting `neighbors` a smaller number than `env_neighbors` wouldn't make it error.

@Charlottez112 `neighbors` and `env_neighbors` are supposed to be provided a `NeighborList` object, or a dict of [query arguments](https://freud.readthedocs.io/en/latest/topics/querying.html#query-arguments). Did freud allow you to pass a numeric value, or do you mean something like "smaller cutoff" or "fewer nearest neighbors"? I get the following error when providing a numerical value:
```python
ValueError: An invalid value was provided for neighbors, which must be a dict or NeighborList object.
```
See [EnvironmentCluster docs](https://freud.readthedocs.io/en/latest/modules/environment.html#freud.environment.EnvironmentCluster) for reference.

**Comment by Charlottez112** [comment]
> > I'm not entirely sure what `neighbors` in the arguments mean. Based on the documentation, it seems like this should always be greater than or equal to `env_neighbors`, but I found setting `neighbors` a smaller number than `env_neighbors` wouldn't make it error.
> 
> @Charlottez112 `neighbors` and `env_neighbors` are supposed to be provided a `NeighborList` object, or a dict of [query arguments](https://freud.readthedocs.io/en/latest/topics/querying.html#query-arguments). Did freud allow you to pass a numeric value, or do you mean something like "smaller cutoff" or "fewer nearest neighbors"? I get the following error when providing a numerical value:
> 
> ```python
> ValueError: An invalid value was provided for neighbors, which must be a dict or NeighborList object.
> ```
I meant e.g. `neighbors={'num_neighbors': 8}` and `env_neighbors={'num_neighbors': 16}`

**Comment by erteich** [comment]
Hi all, I'd be happy to try to help here. The API has changed a lot from when I originally wrote this module, but the way I originally wrote it, env_neighbors should describe the environments of each particle, to be compared, and neighbors should describe which particle neighborhoods are compared. So in the original version of the code, env_neighbors was an integer: the number of nearest neighbors that defines particle i's environment. Then neighbors was also an integer: the number of nearest neighbors to particle i whose environments one seeks to check for similarity with particle i's environment. For example, the diamond structure can be described by two particle environments, each consisting of four nearest neighbors, which are flipped with respect to each other. If one wanted to find all of the particles with type A environments, one could give env_neighbors=4, neighbors=16. This way, particle environments of only 4 nearest neighbors could be compared with other environments that were further away. I have some example scripts I have written to test these types of features, that I could port into the new API, with someone's help, if people would find that useful. I also explained the environment detection method and a lot of its use cases that did not get published in my thesis, haha, which was probably not the most useful or visible place for that information to be- but my thesis is here: 

https://deepblue.lib.umich.edu/handle/2027.42/147721

And in Section 6.1 I discuss some examples and go over how to detect high pressure lithium environments, for example, which are a little more complicated. I would be very happy to work on putting this type of information somewhere a lot more visible and useful, if there was interest or a need for it!

**Comment by Charlottez112** [comment]
Thank you so much @erteich! Your example is super helpful. I'd be happy to work with you to put these information in the documentation. I also have some other questions related to this. I'll follow up with an email to set up a meeting next week if that's okay with you?

**Comment by erteich** [comment]
That sounds great @Charlottez112 ! I'd love to chat next week- my email is erteich@seas.upenn.edu. I look forward to it!

**Comment by vyasr** [comment]
@erteich thanks a bunch for helping out here! Feel free to ping me if any of the API changes cause you problems. FWIW I don't think too much has changed (assuming you have at some point in the last few years familiarized yourself with the way that neighbor searching changed in 2.0 🙂). From a Python perspective IIRC I basically just split the functionality of the original class you wrote into three separate ones so that there wouldn't be any ambiguity about which properties were populated by what functions and what exactly they mean. I believe that in the original implementation some of the computed properties had slightly different interpretations depending on which compute methods was the last one invoked. Now those are all clarified in separate classes.

PR #489 contains those changes, it's a pretty large one but in case it helps looking at just the diff of environment.pyx might help clarify that.

**Comment by Charlottez112** [comment]
Thanks to @erteich now I'm much more comfortable with this module. I wrote a tutorial/example analysis and created a PR in the freud-example repo.
I think we will also need to change the documentation a little bit. For example, we should probably mention that 
1) If `global_search` is set to `True`,  `neighbor` is ignored
2) The `cluster_environments` may have different lengths (may contain 0 vectors)

**Comment by erteich** [comment]
Thanks so much for doing that @Charlottez112 ! Indeed- that global_search flag is now a little bit unnecessary, and could probably be removed if we wanted. We could force both neighbors and env_neighbors to be provided by the user, to remove any ambiguity, at the expense of a little bit of convenience for the user. Previously, I wrote the code such that, if env_neighbors was not provided, it would default to the environments defined by "neighbors." And if "global_search" was True, then all particle neighborhoods would be compared with all other particle neighborhoods, using the environments as defined either by the "neighbors" or "env_neighbors" arguments, depending on what was provided by the user. We could get rid of all these opaque default behaviors though if we just forced the user to provide both neighbors and env_neighbors and dropped the global_search flag.

**Comment by erteich** [comment]
Perhaps we could have a zoom meeting with @vyasr to discuss this? I also was hoping to ask you, Vyas, about some functions in the module in freud v0, like minRMSD and so on, that have been removed in the newer freud. Perhaps there could be a way to reintroduce these functions? I'd love to discuss with you!

**Comment by vyasr** [comment]
Yes sure, happy to get on a call! Are you two free later this week, maybe Friday? @Charlottez112 could you set up a Zoom?
